### DISCUSSION TITLE
How do i use the `DataController` module to make my own Settings menu for my game

### DISCUSSION THREAD

**Discussion by Wilker-uwu** [Ark2000/PankuConsole#158] [Q&A] [answered]
i've been looking for an easier solution to implement [graphics settings](https://www.youtube.com/watch?v=SAH1pXlcN1I "youtube link: display settings demo"), and my current solution involves a lot of jank among the lines of
1. keep a dictionary in the project settings, with an alias and the respective project setting as the key-value pair
2. manually include the buttons and controls in the panel from the ground-up and connect each one to update both the project setting using the respective connected alias
3. write the settings into an override.cfg and use the project settings change event to gather anything that is updated

but because there is either no distinction for which settings to update, the SettingsChanged event is connected to _everything_ that listen to the settings throughout the entire game. the alternative as-is, is to create a signal for every single setting there is and make a mess out of the signals i'll have to connect.

so recently i found out about this addon, but i don't have much of an idea as to how to modify things for my own use case (and i would also like to keep some features of the current visuals that i have, so i'd like to ask here what i can do to have a Class or a Script that can be used to get those settings into my own themed panel that is already built, just like pictured in one of the demo images of this addon.

![](https://github.com/Ark2000/PankuConsole/raw/master/docs/assets/data_controller.png)

thank you for you time <3

**Comment by Ark2000** [comment, marked as answer]
> so recently i found out about this addon, but i don't have much of an idea as to how to modify things for my own use case (and i would also like to keep some features of the current visuals that i have, so i'd like to ask here what i can do to have a Class or a Script that can be used to get those settings into my own themed panel that is already built, just like pictured in one of the demo images of this addon.

The UI auto-generation in Panku Console is deeply coupled with the UI itself, so there's no 'plug and play' method to make your own stylized game menu. However, since Panku Console is designed into many small modules and the `DataController` itself is simply a module, you can check out `addons/panku_console/modules/data_controller` for its implementation and modify for your own needs.

> 1. keep a dictionary in the project settings, with an alias and the respective project setting as the key-value pair
> 2. manually include the buttons and controls in the panel from the ground-up and connect each one to update both the project setting using the respective connected alias
> 3. write the settings into an override.cfg and use the project settings change event to gather anything that is updated

For persistent data storage, a few suggestions:

1. Try to use a lazy update scheme. i.e. add an Apply button to the settings menu, only apply the settings and save it to files when the button is pressed, so you don't have to connect a lot of signals.
2. Just use something like a `.res` or `.json` file to manually store data, do not bother with project settings, they are not for games but mainly for tools imo.

Good luck! I hope it goes well.

**Reply by Wilker-uwu** [reply]
i see! i'll look into it, thank you <3

**Reply by Wilker-uwu** [reply]
hey, it's been about a week, and i still can't quite understand why the component is split-up so much, and there aren't many comments to make clues about too. could you explain to me a bit how is the connection made between the Control and the actual variable update? i think i can go from there if i can grasp this part ^^'

thank you <3

**Reply by Ark2000** [reply]
For exmple:

![image](https://github.com/Ark2000/PankuConsole/assets/21126555/a890ec56-f14a-4cc7-8bb2-8b485f808aff)


1. You click a button to increase an int value, in `row_ui_int.gd`, it emits a signal called `ui_val_changed` with the new value

```gdscript
func _ready():
	line_edit.text_submitted.connect(
		func(_s): 
			ui_val_changed.emit(get_ui_val())
	)
	line_edit.focus_exited.connect(
		func(): 
			ui_val_changed.emit(get_ui_val())
	)
	button_inc.pressed.connect(
		func():
			update_ui(get_ui_val() + 1)
			ui_val_changed.emit(get_ui_val())
	)
	button_dec.pressed.connect(
		func():
			update_ui(get_ui_val() - 1)
			ui_val_changed.emit(get_ui_val())
	)
```

2. in `exporter2.gd`, the `ui_val_changed` signal is captured by a lambda function which will update the actual value in an object

```gdscript
row.ui_val_changed.connect(
	func(val):
		if !is_instance_valid(obj):
			return
		if prop_name in obj:
# call update value function in the object.
			if obj.has_method("update_setting"):
				obj.update_setting(prop_name, val)
			else:
				obj.set(prop_name, val)
)
```

**Reply by Wilker-uwu** [reply]
i see! thank you for the help, i appreciate it <3
